tests/proptest.rs

```diff
@@ -2,35 +2,39 @@
 
 use std::{collections::BTreeSet as Set, error::Error};
 
+use pubgrub::error::PubGrubError;
+use pubgrub::package::Package;
+use pubgrub::range::Range;
+use pubgrub::report::{DefaultStringReporter, Reporter};
+use pubgrub::solver::{
+    choose_package_with_fewest_versions, resolve, Dependencies, DependencyProvider,
+    OfflineDependencyProvider,
+};
+use pubgrub::version::{NumberVersion, Version};
+
 use proptest::collection::{btree_map, vec};
 use proptest::prelude::*;
 use proptest::sample::Index;
 use proptest::string::string_regex;
 
-use pubgrub::range::Range;
-use pubgrub::solver::{resolve, Dependencies, DependencyProvider, OfflineDependencyProvider};
-use pubgrub::{
-    error::PubGrubError, package::Package, report::DefaultStringReporter, report::Reporter,
-    version::NumberVersion, version::Version,
-};
-
 use crate::sat_dependency_provider::SatResolve;
 
 mod sat_dependency_provider;
 
-/// The same as DP but takes versions from the opposite end:
-/// if DP returns versions from newest to oldest, this returns them from oldest to newest.
+/// The same as [OfflineDependencyProvider] but takes versions from the opposite end:
+/// if [OfflineDependencyProvider] returns versions from newest to oldest, this returns them from oldest to newest.
 #[derive(Clone)]
-struct ReverseDependencyProvider<DP>(DP);
-
-impl<P: Package, V: Version, DP: DependencyProvider<P, V>> DependencyProvider<P, V>
-    for ReverseDependencyProvider<DP>
-{
-    fn list_available_versions(&self, p: &P) -> Result<Vec<V>, Box<dyn Error>> {
-        self.0.list_available_versions(p).map(|mut v| {
-            v.reverse();
-            v
-        })
+struct OldestVersionsDependencyProvider<P: Package, V: Version>(OfflineDependencyProvider<P, V>);
+
+impl<P: Package, V: Version> DependencyProvider<P, V> for OldestVersionsDependencyProvider<P, V> {
+    fn make_decision<T: std::borrow::Borrow<P>, U: std::borrow::Borrow<Range<V>>>(
```

Comment:
What about naming this `[pick/choose]_next_package`?

Reply:
Could have been possible. `make_decision` fits well with PubGrub terminology since this step is called ["decision making"](https://github.com/dart-lang/pub/blob/master/doc/solver.md#decision-making)

Reply:
That's true. The question is who do we optimize the easiness to understand for:
1) library authors
2) users

I think `make_decision` optimizes for us, while naming function based on what it does optimizes for users.

What if we use descriptive function name from the user perspective and put a link to "decision making" in the function docs?

Reply:
Yeah I don't know, tough naming call. If you prefer explicitness from the user point of view, maybe `choose_package_version_within` is the most descriptive name? I don't mind if you go for it, but can you also update the guide then please (PR #45).

Reply:
https://github.com/dart-lang/pub/blob/master/lib/src/solver/version_solver.dart#L320 :smile:

Reply:
I went with `choose_package_version`; how does it look now?

I haven't yet read the guide as I'm focusing on PR reviews first. Will update once I get there :slightly_smiling_face: